### ISSUE TITLE
Automatically detect lingering server calls

### ISSUE THREAD

**Issue by leventov** [grpc/grpc-java#6588]
### Problem
When subscribing to a server streaming call, it's tedious to set up machinery for canceling the stream when no longer needed. 

### Describe the solution you'd like
When a user calls `stub.foo(observer)`, gRPC attaches a [Cleaner](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/Cleaner.html) to it which cancels the corresponding `ClientCallStreamObserver` when the inbound stream observer object becomes unreachable.

Since there is some runtime cost of registering Cleaners for each and every streaming call (and there can be thousands of them at any given time, and an unlimited number in the app lifetime), this may best be an opt-in feature. Not sure about exact API, and whether it should be opt-in per-app, per-method, or per-call.

Adding a cleaner also holds a ref to `ClientCallStreamObserver` as long as the inbound stream observer object is alive in the heap, even if the call itself has completed. We need to make sure this doesn't leak significant resources.

### Describe alternatives you've considered
Per-app opt-in is a form of "debug mode" where lingering connections are logged, and then in prod the machinery may be set-up manually in strategic places.

**Comment by voidzcy** [comment]
`Cleaner` is a Java 9 feature, but currently we are supporting a minimal version of Java 7.

**Comment by leventov** [comment]
At least in "debug" mode which is not to be used in production, cleaners could be reasonably polyfilled with WeakReferences + a queue + a low priority thread which loops the queue.

**Comment by ejona86** [comment]
I don't think we would cancel the call for you; that just leads to buggy programs that are relying on the GC for timely behavior. This could be a debug tool and grpc could report when you leak an RPC, like we do for ManagedChannel.
